gax-java/gax/src/main/java/com/google/api/gax/logging/Slf4jUtils.java

```diff
@@ -0,0 +1,268 @@
+    switch (level) {
+      case TRACE:
+        loggingEventBuilder = logger.atTrace();
+        break;
+      case DEBUG:
+        loggingEventBuilder = logger.atDebug();
+        break;
+      case INFO:
+        loggingEventBuilder = logger.atInfo();
+        break;
+      case WARN:
+        loggingEventBuilder = logger.atWarn();
+        break;
+      case ERROR:
+        loggingEventBuilder = logger.atError();
+        break;
+      default:
+        loggingEventBuilder = logger.atDebug();
+        // Default to DEBUG level
+    }
+    contextMap.forEach(loggingEventBuilder::addKeyValue);
+    loggingEventBuilder.log(message);
+  }
+
+  interface LoggerFactoryProvider {
+    ILoggerFactory getLoggerFactory();
+  }
+
+  static class DefaultLoggerFactoryProvider implements LoggerFactoryProvider {
+    @Override
+    public ILoggerFactory getLoggerFactory() {
+      return LoggerFactory.getILoggerFactory();
+    }
+  }
+
+  // logging helper methods
+  static Map<String, Object> messageToMapWithGson(Message message)
+      throws InvalidProtocolBufferException {
+    String json = JsonFormat.printer().print(message);
+    return gson.fromJson(json, new TypeToken<Map<String, Object>>() {}.getType());
+  }
+
+  static void recordServiceRpcAndRequestHeaders(
+      String serviceName,
+      String rpcName,
+      String endpoint,
+      Map<String, String> requestHeaders,
+      LogData.Builder logDataBuilder,
+      LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isInfoEnabled()) {
+            addIfNotEmpty(logDataBuilder::serviceName, serviceName);
+            addIfNotEmpty(logDataBuilder::rpcName, rpcName);
+            addIfNotEmpty(logDataBuilder::httpUrl, endpoint);
+          }
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.requestHeaders(requestHeaders);
+          }
+        });
+  }
+
+  private static void addIfNotEmpty(Consumer<String> setter, String value) {
+    if (value != null && !value.isEmpty()) {
+      setter.accept(value);
+    }
+  }
+
+  static void recordResponseHeaders(
+      Map<String, String> headers, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.responseHeaders(headers);
+          }
+        });
+  }
+
+  static <RespT> void recordResponsePayload(
+      RespT message, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            Map<String, Object> messageToMapWithGson =
+                Slf4jUtils.messageToMapWithGson((Message) message);
+
+            logDataBuilder.responsePayload(messageToMapWithGson);
```

Comment:
qq, how does this work for streaming calls where it's possible to receive multiple payloads per RPC call?

Would each payload be overwritten? Or are the payload chained together?

Reply:
as discussed, will create a followup ticket to verify/test. Add showcase tests.

Reply:
created https://github.com/googleapis/sdk-platform-java/issues/3645 and https://github.com/googleapis/sdk-platform-java/issues/3646